app/helpers/textual_summary_helper.rb

```diff
@@ -1,4 +1,8 @@
 module TextualSummaryHelper
+  def role_allows?(_options = {})
+    raise NotImplementedError, _("role_allows? must be implemented in including class")
+  end
```

Comment:
This is super weird :confused: I'm not sure if we want to go this way :confused:

Reply:
If you look through the core code you will see this pattern implemented in several places. Though, admittedly, these are in classes not modules.

What you have with TextualSummaryHelper is a module that is dependent, at least in part, on certain other methods being defined. Typically in the programming world that would be called an interface, i.e. a collection of methods that must be defined.

Ruby doesn't have real interfaces, hence we have to resort to this pattern. See also https://github.com/ManageIQ/manageiq/issues/19431

Reply:
My concern is going fully Java here :disappointed:

Reply:
> My concern is going fully Java here

I understand, and in small projects that pattern is usually not necessary. But in very big projects (of which this is one IMO) those types of patterns start to become essential, or you end up with a bunch of hidden design quirks like this. And, so far as I can tell (in our code anyway), it's almost never documented. So, this pattern does double duty in that regard - it defines it *and* it's self-documenting if an error does occur.